sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuWindowGroupLimitExec.scala

```diff
@@ -0,0 +1,263 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "350"}
+{"spark": "351"}
+spark-rapids-shim-json-lines ***/
+package com.nvidia.spark.rapids.shims
+
+import ai.rapids.cudf
+import ai.rapids.cudf.GroupByOptions
+import com.nvidia.spark.rapids.{BaseExprMeta, DataFromReplacementRule, GpuBindReferences, GpuBoundReference, GpuColumnVector, GpuExec, GpuExpression, GpuMetric, GpuOverrides, GpuProjectExec, RapidsConf, RapidsMeta, SparkPlanMeta, SpillableColumnarBatch, SpillPriorities}
+import com.nvidia.spark.rapids.Arm.withResource
+import com.nvidia.spark.rapids.RmmRapidsRetryIterator.{splitSpillableInHalfByRows, withRetry}
+import com.nvidia.spark.rapids.window.{GpuDenseRank, GpuRank}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, DenseRank, Expression, Rank, SortOrder}
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.window.{WindowGroupLimitExec, WindowGroupLimitMode}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+sealed trait RankFunctionType
+case object RankFunction extends RankFunctionType
+case object DenseRankFunction extends RankFunctionType
+
+class GpuWindowGroupLimitExecMeta(limitExec: WindowGroupLimitExec,
+                                  conf: RapidsConf,
+                                  parent:Option[RapidsMeta[_, _, _]],
+                                  rule: DataFromReplacementRule)
+    extends SparkPlanMeta[WindowGroupLimitExec](limitExec, conf, parent, rule) {
+
+  private val partitionSpec: Seq[BaseExprMeta[Expression]] =
+    limitExec.partitionSpec.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+
+  private val orderSpec: Seq[BaseExprMeta[SortOrder]] =
+    limitExec.orderSpec.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+
+  private val rankFunction = GpuOverrides.wrapExpr(limitExec.rankLikeFunction, conf, Some(this))
+
+  override def tagPlanForGpu(): Unit = {
+    wrapped.rankLikeFunction match {
+      case DenseRank(_) =>
+      case Rank(_) =>
+      // case RowNumber() => // TODO: Future.
+      case _ => willNotWorkOnGpu("Only Rank() and DenseRank() are " +
+                                 "currently supported for window group limits")
+    }
+  }
+
+
+  override def convertToGpu(): GpuExec = {
+    GpuWindowGroupLimitExec(partitionSpec.map(_.convertToGpu()),
+                            orderSpec.map(_.convertToGpu().asInstanceOf[SortOrder]),
+                            rankFunction.convertToGpu(),
+                            limitExec.limit,
+                            limitExec.mode,
```

Comment:
Can we please validate that this is one of Partial or Final?  Just being paranoid here.

Reply:
Done.